### PR TITLE
Optimizing grpc_error Part 1

### PR DESCRIPTION
**Pull request by ncteisen**
Based on #9745

This is the first of a few PRs that will improve grpc_error performance. This improvement checks if an error passed to copy_error_and_unref is unique. If so, it simply returns the old error instead of allocating a new one.

### CODE REVIEW

File: src/core/lib/iomgr/error.c [hunk 1]
```diff
@@ -249,6 +249,9 @@ grpc_error *grpc_error_create(const char *file, int line, const char *desc,
 
 static grpc_error *copy_error_and_unref(grpc_error *in) {
   GPR_TIMER_BEGIN("copy_error_and_unref", 0);
+  if (gpr_ref_is_unique(&in->refs)) {
```
Comment (ctiller) [inline src/core/lib/iomgr/error.c]: This should be moved down to line 269 (maybe as an `else if` on the main branch path): if `in` is special, then we'll segfault if we try to refer to `refs`.

Comment (ncteisen) [inline src/core/lib/iomgr/error.c]: Oops! Done.

**Comment by ctiller** [comment]
ASAN is showing some leaks in error_test

**Comment by ncteisen** [comment]
ASAN is happier now that I remembered to clean up after myself.

Also, as a note. I decided to make this the branch point for the upcoming arena optimization. Figured it would be easier to redo inlining and slice after arena, than to try to implement arena with my other changes in place

**Comment by ctiller** [comment]
Looks like clang-format is unhappy

`tools/distrib/clang_format_code.sh`

**Comment by ncteisen** [comment]
**Failures:**

Basic Mac: #9756 
Interop #9937 

Linux:
- c_linux_dbg_native.bins/dbg/h2_ssl_proxy_test no_logging GRPC_POLL_STRATEGY=epoll
- python_linux_opt_native.py27.test.protoc_plugin.beta_python_plugin_test.PythonPluginTest
- aggregate_tests.run_tests_python_linux_opt_native
- aggregate_tests.run_tests_c_linux_dbg_native

The linux ones aren't tracked, but don't seem related

**Comment by ncteisen** [comment]
Same failures occur as mentioned in previous comment.

Additionally, Linux's End2endTest.SimultaneousReadWritesDone.

**Pull request merged**